backend-comparison/src/burnbenchapp/auth.rs

```diff
@@ -0,0 +1,191 @@
+use reqwest;
+use std::error::Error;
+use std::io::Write;
+#[cfg(unix)]
+use std::os::unix::fs::PermissionsExt;
+use std::{
+    fs::{self, File},
+    path::{Path, PathBuf},
+};
+
+pub(crate) static CLIENT_ID: &str = "Iv1.84002254a02791f3";
+static GITHUB_API_VERSION_HEADER: &str = "X-GitHub-Api-Version";
+static GITHUB_API_VERSION: &str = "2022-11-28";
+
+/// Return the file path for the auth cache on disk
+pub fn get_auth_cache_file_path() -> PathBuf {
+    let home_dir = dirs::home_dir().expect("an home directory should exist");
+    let path_dir = home_dir.join(".cache").join("burn").join("burnbench");
+    #[cfg(test)]
+    let path_dir = path_dir.join("test");
+    let path = Path::new(&path_dir);
+    path.join("token.txt")
+}
+
+/// Return true if the token is still valid
+pub(crate) fn is_token_valid(token: &str) -> bool {
+    get_username_from_token(token).is_ok()
+}
+
+/// Retrieve the user name from the access token
+fn get_username_from_token(token: &str) -> Result<String, Box<dyn Error>> {
+    let client = reqwest::blocking::Client::new();
+    // User-Agent is important otherwise GitHub rejects the request with a 403
+    // See: https://github.com/seanmonstar/reqwest/issues/918#issuecomment-632906966
+    let response = client
+        .get("https://api.github.com/user")
+        .header(reqwest::header::USER_AGENT, "burnbench")
+        .header(reqwest::header::ACCEPT, "application/vnd.github+json")
+        .header(reqwest::header::AUTHORIZATION, format!("Bearer {}", token))
+        .header(GITHUB_API_VERSION_HEADER, GITHUB_API_VERSION)
+        .send()?;
+    let response = response.json::<serde_json::Value>()?;
+    let username = response["login"].as_str().map(|s| s.to_string());
+    // Return an error if the login field is not found
+    username.ok_or_else(|| {
+        From::from(std::io::Error::new(
+            std::io::ErrorKind::NotFound,
+            "Username not found in the response",
+        ))
+    })
+}
```

Comment:
Can we avoid to allocate the `String` here and use lifetimes in order to retrieve the `bool`? 

`get_username_from_token` seems not used anywhere in the code

Reply:
I don't think using liketime here works, since the lifetime of the `str` is associated with the response in JSON from the request that will be cleaned up after the function, so String is all right.

Reply:
> get_username_from_token seems not used anywhere in the code

Well it is used in `is_token_valid` 😄
We could make it inline in `is_token_valid` but we would loose some meaning.

Reply:
@syl20bnr 

Yep, your analysis is correct. You can solve this returning directly the `bool`  in `get_username_from_token` or insert `get_username_from_token` into `is_token_valid`.  You do not catch the `Result` at all in that way because you are using `ok` so the error is never unwrapped.

Reply:
Got it, thank you. And I see your point. I am not sure yet that I don't want the username at the CLI tool level so I prefer to keep it now as this even if it is not really used.

Reply:
A possible tradeoff could be to convert `get_username_from_token` in a closure.  If I recall correctly the compiler treats them differently from a function when it produces the code, without bringing into play the calling stack.

Reply:
I replaced this solution with `verify_token` that just checks for the status code of the request.